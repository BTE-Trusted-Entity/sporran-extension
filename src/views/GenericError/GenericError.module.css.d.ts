export const __esModule: true;
export const overlay: string;
export const text: string;
export const details: string;
