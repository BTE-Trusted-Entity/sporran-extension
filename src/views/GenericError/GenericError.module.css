.overlay {
  composes: overlay from '../../components/Overlay/Overlay.module.css';
  max-width: 100%;
  max-height: 100%;
  height: calc(100% - 2rem);
  inset: 1rem;
  padding: 1rem;
}

.text {
  composes: text from '../../components/Overlay/Overlay.module.css';
  background: url('../../components/images/transaction-error.svg') no-repeat top
    center/4rem;
  padding-top: 5rem;
  margin-top: 0;
  margin-bottom: 1rem;
}

.text a {
  color: var(--color-lightest);
}

.details {
  composes: input from '../../components/Input/Input.module.css';
  flex-grow: 1;
  align-self: stretch;
  margin-bottom: 1rem;
}
