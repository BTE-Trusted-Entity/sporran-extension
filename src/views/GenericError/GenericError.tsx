import { Component } from 'react';
import { Modal } from 'react-dialog-polyfill';
import { browser } from 'webextension-polyfill-ts';

import * as overlayStyles from '../../components/Overlay/Overlay.module.css';
import * as styles from './GenericError.module.css';

import { configuration } from '../../configuration/configuration';

interface Props {
  children: JSX.Element;
}

interface State {
  errorText?: string;
}

export class GenericError extends Component<Props, State> {
  state: State = {};

  static getDerivedStateFromError(error: Error): State {
    return {
      errorText: `Sporran@${configuration.version}\n\n${error.message}\n\n${error.stack}`,
    };
  }

  render(): JSX.Element {
    if (!this.state.errorText) {
      return this.props.children;
    }

    const t = browser.i18n.getMessage;

    return (
      <main>
        <Modal open className={styles.overlay}>
          <p
            className={styles.text}
            dangerouslySetInnerHTML={{ __html: t('view_GenericError_message') }}
          />

          <textarea
            className={styles.details}
            readOnly
            aria-label={t('view_GenericError_details')}
            defaultValue={this.state.errorText}
            onFocus={(event) => {
              event.target.select();
            }}
          />

          <button
            type="button"
            onClick={() => window.close()}
            className={overlayStyles.button}
          >
            {t('common_action_close')}
          </button>
        </Modal>
      </main>
    );
  }
}
