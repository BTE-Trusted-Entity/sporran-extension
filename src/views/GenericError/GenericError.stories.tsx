import { Meta } from '@storybook/react';
import { useState, MouseEvent } from 'react';

import { GenericError } from './GenericError';

export default {
  title: 'Views/GenericError',
  component: GenericError,
} as Meta;

function Failure(): JSX.Element {
  const [error, setError] = useState<MouseEvent | null>(null);

  if (error) {
    throw new Error('Render failure triggered');
  }

  return (
    <button type="button" onClick={setError}>
      Trigger error
    </button>
  );
}

export function Template(): JSX.Element {
  return (
    <GenericError>
      <Failure />
    </GenericError>
  );
}
