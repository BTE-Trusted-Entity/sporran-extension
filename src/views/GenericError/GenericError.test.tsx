import {
  mockDialogShowModal,
  render,
  waitForDialogUpdate,
} from '../../testing/testing';

import { GenericError } from './GenericError';

describe('GenericError', () => {
  beforeEach(mockDialogShowModal);

  it('should render children when no error happens', async () => {
    const { container } = render(
      <GenericError>
        <span>Pass</span>
      </GenericError>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render generic error when an error happens', async () => {
    function Failure(): JSX.Element {
      const error = new Error('Testing GenericError');
      error.stack = 'test.js@123:12';
      throw error;
    }

    const logger = jest.spyOn(console, 'error');
    logger.mockImplementation(() => undefined);

    const { container } = render(
      <GenericError>
        <Failure />
      </GenericError>,
    );

    await waitForDialogUpdate();
    expect(container).toMatchSnapshot();

    logger.mockRestore();
  });
});
