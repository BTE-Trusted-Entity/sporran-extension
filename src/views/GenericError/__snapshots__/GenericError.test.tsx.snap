// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericError should render children when no error happens 1`] = `
<div>
  <span>
    Pass
  </span>
</div>
`;

exports[`GenericError should render generic error when an error happens 1`] = `
<div>
  <main>
    <dialog
      class="overlay"
    >
      <p
        class="text"
      >
        An error occurred; please try again. For help please contact  
        <a
          href="mailto:tech-support@sporran.org"
        >
          tech-support@sporran.org
        </a>
         and send over the following text in full.
      </p>
      <textarea
        aria-label="Error details"
        class="details"
        readonly=""
      >
        Sporran@2023.4.12

Testing GenericError

test.js@123:12
      </textarea>
      <button
        class="button"
        type="button"
      >
        Close
      </button>
    </dialog>
  </main>
</div>
`;
